samza-core/src/main/java/org/apache/samza/zk/ZkUtils.java

```diff
@@ -215,14 +235,62 @@ public void close() throws ZkInterruptedException {
     zkClient.close();
   }
 
+  // Generation enforcing zk listener abstract class.
+  // Helps listeners, which extend it,  to skip old generation events.
+  // We cannot use 'sessionId' for this because it is not available through ZkClient (at leaste without reflection)
+  public abstract static class GenIZkChildListener implements IZkChildListener {
+    private final int generation;
+    private final ZkUtils zkUtils;
+
+    public GenIZkChildListener(ZkUtils zkUtils) {
+      generation = zkUtils.getGeneration();
+      this.zkUtils = zkUtils;
+    }
+
+    protected boolean skip(String listenerName) {
+      int curGeneration = zkUtils.getGeneration();
+      if (curGeneration != generation) {
+        LOG.warn("SKIPPING handleDataChanged for " + listenerName +
+            " from wrong generation. current generation=" + curGeneration + "; callback generation= " + generation);
+        return true;
+      }
+      return false;
+    }
+  }
+
+  public abstract static class GenIZkDataListener implements IZkDataListener {
+    private final int generation;
+    private final ZkUtils zkUtils;
+    private final String listenerName;
+
+    public GenIZkDataListener(ZkUtils zkUtils, String listenerName) {
+      generation = zkUtils.getGeneration();
+      this.zkUtils = zkUtils;
+      this.listenerName = listenerName;
+    }
+
+    protected boolean notAValidEven() {
```

Comment:
typo in method name: s/notAValidEven/notAValidEvent/

in the previous listener, you have used the method name as skip(). Does it make sense to use the same for both?